src/helics/core/mpi/MpiService.cpp

```diff
@@ -0,0 +1,314 @@
+{
+    for (unsigned int i = 0; i < comms.size (); i++)
+    {
+        if (comms[i] == comm)
+        {
+            // If somehow this gets called while MPI is still initializing, wait until MPI initialization completes
+            while (startup_flag && !stop_service);
+
+            return std::to_string(commRank) + ":" + std::to_string(i);
+        }
+    }
+
+    // Comm not found
+    return "";
+}
+
+int MpiService::getRank ()
+{
+    // If somehow this gets called while MPI is still initializing, wait until MPI initialization completes
+    while (startup_flag && !stop_service);
+
+    return commRank;
+}
+
+
+int MpiService::getTag (MpiComms *comm)
+{
+    for (unsigned int i = 0; i < comms.size (); i++)
+    {
+        if (comms[i] == comm)
+        {
+            return i;
+        }
+    }
+
+    return -1;
+}
+
+bool MpiService::initMPI()
+{
+    // Initialize MPI with MPI_THREAD_FUNNELED
+    int mpi_initialized;
+    int mpi_thread_level;
+
+    MPI_Initialized(&mpi_initialized);
+
+    if (!mpi_initialized)
+    {
+        MPI_Init_thread(nullptr, nullptr, MPI_THREAD_FUNNELED, &mpi_thread_level);
+
+        MPI_Initialized(&mpi_initialized);
+        if (!mpi_initialized)
+        {
+            std::cerr << "MPI initialization failed" << std::endl;
+            return false;
+        }
+    }
+
+    MPI_Query_thread(&mpi_thread_level);
+
+    if (mpi_thread_level < MPI_THREAD_FUNNELED)
+    {
+        std::cerr << "MPI_THREAD_FUNNELED support required" << std::endl;
+        return false;
+    }
+
+    return true;
+}
+
+void MpiService::sendAndReceiveMessages ()
+{
+    // Eventually this entire loop should be redone
+    // Using fixed size chunks for sending messages would allow posting blocks of irecv requests
+    // If we know that a message will get received, a blocking MPI_Wait_any could be used for send requests
+    // Also, a method of doing time synchronization using MPI reductions should be added
+    for (unsigned int i = 0; i < comms.size(); i++)
+    {
+        // Skip any nullptr entries
+        if (comms[i] == nullptr)
+        {
+            continue;
+        }
+
+        // Handle receives for the MpiComms object
+        int message_waiting = 1;
+        MPI_Status status;
+
+        while (message_waiting)
+        {
+            // Check if there is a messages waiting
+            MPI_Iprobe (MPI_ANY_SOURCE, i, mpiCommunicator, &message_waiting, &status);
+
+            if (message_waiting)
+            {
+                // Get the size of the message waiting to be received
+                int recv_size;
+                std::vector<char> buffer;
+                MPI_Get_count (&status, MPI_CHAR, &recv_size);
+                buffer.resize (recv_size);
+
+                // Post an asynchronous receive
+                MPI_Request req;
+                MPI_Irecv (buffer.data (), buffer.capacity (), MPI_CHAR, MPI_ANY_SOURCE, i, mpiCommunicator, &req);
+
```

Comment:
You're using a non-blocking receive in a blocking fashion, since you are spinning on the MPI_Test later.  You could consider using a blocking receive instead unless there is other work to do while spinning on the request.

Reply:
Originally the receive was happening in a separate thread from the sends and to serialize the calls it had to be non-blocking. I was thinking of sticking the receive request in an array with the buffer to be filled, then later on check for finished receives.